chore(broker): split snapshot replication destination folders

npepinpe opened the pull request:
## Description

- splits destination folders for the different snapshot replications as they run in different thread context and both need to modify the same on-disk state
- Atomix snapshot replication will now write to raft-pending/ folder
- Zeebe snapshot replication will now write to pushed-pending/ folder

## Related issues

closes #3531, #4078, #3887, #3764

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

npepinpe commented:
Will run a benchmark and report later.

deepthidevaki commented on broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java:
```diff
@@ -23,32 +23,37 @@
     this.deletionListeners = new CopyOnWriteArraySet<>();
-    this.store.addListener(this);
 
+    IoUtil.ensureDirectoryExists(pendingDirectory.toFile(), "push replication pending directory");
```
We can use `FileUtil:ensureDirectoryExists` which you added in another PR to be consistent.

npepinpe replied:
As soon as the other PR is merged :+1:

deepthidevaki commented on broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/AtomixSnapshotStorage.java:
```diff
@@ -142,8 +147,13 @@ public void onNewSnapshot(
+    try {
+      LOGGER.debug("Purging snapshots older than {}", snapshot);
+      store.purgeSnapshots(snapshot);
+      DbSnapshotStore.cleanUpTemporarySnapshots(pendingDirectory, snapshot.index());
```
Doesn't purgeSnapshot does the cleanup?

npepinpe replied:
It now only does for the one pending directory it knows about - do you think it should also purge this one as well? I'm also not sure if it's a good idea for it do it anyway, because `purgeSnapshots` can be called from different thread contexts, so we might still run into the issue of deleting on going replications? Though it should be more fault tolerant

deepthidevaki replied:
Sorry, I missed that DbSnapshot is shared between raft and atomix. Then it make sense to cleanup here. But then wouldn't we cleanup raft-pending snapshot while we purge snapshots here?

npepinpe commented:
Blocked by https://github.com/zeebe-io/atomix/pull/205

deepthidevaki commented on broker/src/main/java/io/zeebe/broker/clustering/atomix/storage/snapshot/DbSnapshotStore.java:
```diff
@@ -170,6 +171,35 @@ public void removeListener(final SnapshotListener listener) {
+    }
+  }
+
+  static void cleanUpTemporarySnapshots(final Path pendingDirectory, final long cutoffIndex)
```
This isn't used anymore. Can remove.

npepinpe commented:
Thanks! Two things I addressed, and one question:

1. Removed the dead code
1. I moved the creation of the pending directory to the ZeebePartition because I generally dislike having constructors throw things like I/O exceptions or perform I/O, and I felt it's cleaner to fail the installation at that point by return a failed future - let me know what you think. It's not the cleanest now tbh, but I wasn't sure how else to structure it

And last, one question I thought, this is potentially "breaking" as we don't used the older `pending/` directory anymore, potentially leaving partial snapshots. Should we use `pending/` instead of `raft-pending/`? Long term, I would like to merge both replication mechanisms, so probably our only mechanism left would be the raft one, so we would go back to a single folder - might make sense to leave it named `pending/`. I thought it'd be more confusing initially, but now I think it's better :sweat_smile:

npepinpe commented:
I think I broke something rebasing...don't review yet, I'll fix it :sweat_smile:

npepinpe commented:
Also, regarding purging pending snapshots, we have a replication metric which I think may get out of sync. It's currently incremented/decremented entirely by the follower in the `ReplicationController` - increment when we start receiving a new snapshot, decrement if we invalidate it or complete it.

But let's say I start receiving snapshot X, then a new snapshot X + 1 is committed, and we purge X - but then I receive another chunk, so I increment the metric again. And maybe I never receive the last chunk, so I will never decrement it...

Would it be enough to decrement the metric here when we purge? Technically it may get incremented again by the replication controller, but then either it is eventually completed/invalidated or we purge it again and it's decremented.

deepthidevaki commented:
> But let's say I start receiving snapshot X, then a new snapshot X + 1 is committed, and we purge X - but then I receive another chunk, so I increment the metric again. And maybe I never receive the last chunk, so I will never decrement it...

It won't increment the metric because it consider it as the continuation of ongoing replication of X. 

But I'm seeing another potential error here. start receiving snapshot X, then a new snapshot X + 1 is committed, and we purge pending X , received the last chunk of X , commit X because it thinks it has all chunks, but all the previous chunks were deleted :see_no_evil:  This scenario where X+1 commits before X is extremely rare though.

npepinpe commented:
You're right though, that is potentially problematic...

Some possible solutions:

1. Only ensure that the directory exists when a new replication starts, which results in an error if you try to write a chunk after the purge
1. Calculate total snapshot checksum (e.g. checksum of every file so far, not just individual files), and on completion compute the total checksum again from the on-disk state and compare it

I would lean towards the second, but we could do this in another issue as this isn't a bug we're introducing here

npepinpe commented:
bors r+

npepinpe commented:
bors r+

npepinpe commented:
Build got killed :disappointed_relieved: 

bors retry
